keras_nlp/layers/transformer_decoder.py

```diff
@@ -306,26 +309,33 @@ def call(
         x = decoder_sequence  # Intermediate result.
 
         # Compute self attention mask.
-        batch_size = tf.shape(decoder_sequence)[0]
-        input_length = output_length = tf.shape(decoder_sequence)[1]
-        # We need to handle a rectangular causal mask when doing cached
-        # decoding. For generative inference, `decoder_sequence` will
-        # generally be length 1, and `cache` will be the full generation length.
-        if self_attention_cache is not None:
-            input_length = tf.shape(self_attention_cache)[2]
-        self_attention_mask = compute_causal_mask(
-            batch_size,
-            input_length,
-            output_length,
-            0
-            if self_attention_cache_update_index is None
-            else self_attention_cache_update_index,
-        )
         decoder_mask = merge_padding_and_attention_mask(
             decoder_sequence, decoder_padding_mask, decoder_attention_mask
         )
-        if decoder_mask is not None:
-            self_attention_mask = tf.minimum(decoder_mask, self_attention_mask)
+        if use_causal_mask:
+            batch_size = tf.shape(decoder_sequence)[0]
```

Comment:
I think this branching is getting complex enough that we should split a private method for this

```
self_attention_mask = self._compute_self_attention_mask(
    decoder_sequence,
    self_attention_cache,
    self_attention_cache_update_index,
    use_causal_mask,
)
```

Reply:
good point, done